pkg/collector/corechecks/servicediscovery/module/impl_linux.go

```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
 // Register registers the discovery module with the provided HTTP mux.
 func (s *discovery) Register(httpMux *module.Router) error {
 	httpMux.HandleFunc("/status", s.handleStatusEndpoint)
+	httpMux.HandleFunc(patchListeners, s.handleOpenPorts)
 	return nil
 }
 
 // Close cleans resources used by the discovery module.
-// Currently, a no-op.
-func (s *discovery) Close() {}
+func (s *discovery) Close() {
+	clear(s.cache)
+}
 
 // handleStatusEndpoint is the handler for the /status endpoint.
 // Reports the status of the discovery module.
 func (s *discovery) handleStatusEndpoint(w http.ResponseWriter, _ *http.Request) {
 	_, _ = w.Write([]byte("Discovery Module is running"))
 }
+
+func (s *discovery) handleError(w http.ResponseWriter, route string, status int, err error) {
+	_ = log.Errorf("failed to handle /discovery/%s (status: %d): %v", route, status, err)
+	w.WriteHeader(status)
+}
+
+func (s *discovery) handleOpenPorts(w http.ResponseWriter, _ *http.Request) {
+	listeners, err := s.getListeners()
+	if err != nil {
+		s.handleError(w, patchListeners, http.StatusInternalServerError, fmt.Errorf("failed to get listeners: %v", err))
+		return
+	}
+
+	resp := &model.ListenersResponse{
+		Listeners: *listeners,
+	}
+	utils.WriteAsJSON(w, resp)
+}
+
+var allowedEnvironmentVariables = map[string]struct{}{
+	"PATH":                     {},
+	"PWD":                      {},
+	"GUNICORN_CMD_ARGS":        {},
+	"WSGI_APP":                 {},
+	"DD_SERVICE":               {},
+	"DD_TAGS":                  {},
+	"DD_INJECTION_ENABLED":     {},
+	"SPRING_APPLICATION_NAME":  {},
+	"SPRING_CONFIG_LOCATIONS":  {},
+	"SPRING_CONFIG_NAME":       {},
+	"SPRING_PROFILES_ACTIVE":   {},
+	"CORECLR_PROFILER_PATH":    {},
+	"CORECLR_ENABLE_PROFILING": {},
+	"JAVA_TOOL_OPTIONS":        {},
+	"_JAVA_OPTIONS":            {},
+	"JDK_JAVA_OPTIONS":         {},
+	"JAVA_OPTIONS":             {},
+	"CATALINA_OPTS":            {},
+	"JDPA_OPTS":                {},
+	"VIRTUAL_ENV":              {},
+}
```

Comment:
Should be modified to reflect only the environment variables we actually need to save for service-name extraction